### PR TITLE
fixed issue where tests should fail when code has error

### LINKED ISSUE #36000: [learn] tests should fail when code has error
I entered a valid solution and passed all the tests in the picture below. Then I added the `a` on the second line: `var collection = { a` - to make the build fail, but all the test results still have the green check mark like they passed. I think they should go back to failed if there's an error.

Screen Shot 2019-05-08 at 8 54 04 AM

### PR DESCRIPTION
**Pull request by Sfailla**
Signed-off-by: Stevo99 <Sfailla1983@gmail.com>

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #36000

### CODE REVIEW

**Comment by raisedadead** [comment]
Thanks for the PR. Can you update the template above as well?

**Comment by RandellDawson** [comment]
@Stevo99 Is this related to a current issue?  If so, make sure to state "Closes #issueNumber" in the PR instead of just leaving it "Closes #####".

Thanks.

Also, it is not necessary to put "Signed-off-by: Stevo99 Sfailla1983@gmail.com"

**Comment by Stevo99** [comment]
ok thanks!

**Comment by RandellDawson** [comment]
@Stevo99 You have referenced this PR's number instead of an issue number.

**Comment by Stevo99** [comment]
hopefully I got it right now.  I got the number from the initial issue I first responded to.

**Comment by thecodingaviator** [comment]
@Stevo99 it looks good now. Thanks

**Review by ojeytonwilliams — DISMISSED** [review]
So, first off, the code does fix the problems in the issue.  However, I think it should be done a little differently, since it is a bit indirect.  Rather than updating the tests as failing, it tries to execute the tests (which we know will all fail) using the build error as a test runner.  This fails and so the tests are set as failing.  However, at first glance it's hard to understand *why* the code attempts to execute the tests in the catch block, since you should only reach it if something has gone wrong.

Instead, @ValeraS suggested, a better approach would be to start by setting the tests as failing. 
Then they can be passed to `executeTests` if the build passes or straight to `updateTests` if the build fails.  I think the logic is easier to follow this way.

**Comment by Stevo99** [comment]
I'm having trouble starting the tests as failing.  Should I be making the changes in `executeChallengeSaga()` function or somewhere else?

**Comment by ojeytonwilliams** [comment]
Sorry @Stevo99, I completely missed your comment.  I didn't mean to leave you hanging.

> Should I be making the changes in executeChallengeSaga() function or somewhere else?

Mostly `executeChallengeSaga()`, yes. If you get the tests there instead of in `executeTests()`, you can set them as failing (each test needs the property `pass: false`) and pass them to `executeTests()` along with the test runner.  That way if the build fails, the tests still count as failing, even though they're never executed.  Let me know if that works or not.

**Comment by Stevo99** [comment]
No prob I know you guys are busy and I feel bad bothering you lol. Ok I was close let me go back and see if I can get it now thanks!

**Comment by Stevo99** [comment]
I made those changes and it didn't work.  I believe it would work but the logic has to change a little bit in `executeTests()`.  I've been trying to figure it out for the last few days but unfortunately have not been able to.  Let me know if you have any ideas. thanks!

**Comment by ojeytonwilliams** [comment]
How about I have a go at solving it and add a commit to this PR? That way you'll still be credited for the work you've put into this. If you could review the code, since you've been looking at this, that would help with the QA. How does that sound?

**Comment by Stevo99** [comment]
Yea that sounds good to me.  Thanks!

**Comment by ojeytonwilliams** [comment]
@ValeraS could you take a look?

File: client/src/templates/Challenges/redux/execute-challenge-saga.js [hunk 1]
```diff
@@ -36,9 +36,17 @@ export function* executeChallengeSaga() {
...
   const consoleProxy = yield channel();
+  let testResults;
+
   try {
     yield put(initLogs());
     yield put(initConsole('// running tests'));
+    const tests = yield select(challengeTestsSelector);
+    // initialize testResults as failing in case the build fails.
+    testResults = tests.map(test => {
+      test.err = 'build failed';
```
Comment (ValeraS) [inline client/src/templates/Challenges/redux/execute-challenge-saga.js]: IMO, it's better to mark tests as not executed:
```js
	const tests = (yield select(challengeTestsSelector)).map(
      ({ text, testString }) => ({ text, testString })
    );
    yield put(updateTests(tests));
```

File: client/src/templates/Challenges/redux/execute-challenge-saga.js [hunk 2]
```diff
@@ -79,8 +88,7 @@ function* buildChallengeData(challengeData) {
   }
 }
 
-function* executeTests(testRunner) {
-  const tests = yield select(challengeTestsSelector);
+function* executeTests(testRunner, tests) {
```
Comment (ValeraS) [inline client/src/templates/Challenges/redux/execute-challenge-saga.js]:
```suggestion
function* executeTests(testRunner, tests, testTimeout = 5000) {
```

File: client/src/templates/Challenges/redux/execute-challenge-saga.js [hunk 3]
```diff
@@ -51,12 +59,13 @@ export function* executeChallengeSaga() {
       proxyLogger,
       document
     );
-    const testResults = yield executeTests(testRunner);
+    testResults = yield executeTests(testRunner, tests);
```
Comment (ValeraS) [inline client/src/templates/Challenges/redux/execute-challenge-saga.js]:
```suggestion
    const testResults = yield executeTests(testRunner, tests);
```

File: client/src/templates/Challenges/redux/execute-challenge-saga.js [hunk 4]
```diff
@@ -51,12 +59,13 @@ export function* executeChallengeSaga() {
...
 
     yield put(updateTests(testResults));
     yield put(updateConsole('// tests completed'));
     yield put(logsToConsole('// console output'));
   } catch (e) {
+    yield put(updateTests(testResults));
```
Comment (ValeraS) [inline client/src/templates/Challenges/redux/execute-challenge-saga.js]:
```suggestion
```

**Review by ValeraS — CHANGES_REQUESTED** [review]

**Comment by ojeytonwilliams** [comment]
@ValeraS I made the requested changes.  There was a webpack error when I used the syntax you suggested (not sure why, it's fine).  That's why I ended up using a longer expression.

**Review by ValeraS — APPROVED** [review]

**Comment by ojeytonwilliams** [comment]
Huh.  That definitely didn't work on my machine, but it does now.  Not sure what changed.

**Pull request merged**